utils/text_serializer.c

```diff
@@ -17,63 +17,65 @@
 #define UINT32_LEN 10
 #define UINT64_LEN 20
 
-status_t serialize_msg(const uint8_t *iv, uint32_t ciphertext_len, const char *ciphertext, const uint64_t timestamp,
-                       const uint8_t *hmac, char *out_msg, size_t *out_msg_len) {
+status_t serialize_msg(const ta_cipher_ctx *ctx, char *out_msg, size_t *out_msg_len) {
   /* FIXME: Provide some checks here */
   char str_ciphertext_len[UINT32_LEN + 1] = {0};
   char buf[UINT64_LEN + 1] = {0};
   char *ptr = out_msg;
 
-  snprintf(str_ciphertext_len, UINT32_LEN + 1, "%010u", ciphertext_len);
-  // initialize vector
-  if (iv) {
-    memcpy(ptr, iv, IV_LEN);
-  } else {
-    memset(ptr, 0, IV_LEN);
+  if (ctx->ciphertext == NULL || out_msg == NULL || out_msg_len == NULL) {
+    fprintf(stderr, "NULL pointer deferenece\n");
+    return SC_UTILS_TEXT_SERIALIZE;
   }
-  ptr += IV_LEN;
 
+  snprintf(str_ciphertext_len, UINT32_LEN + 1, "%010lu", ctx->ciphertext_len);
+  // initialize vector
+  memcpy(ptr, ctx->iv, IV_LEN);
+  ptr += IV_LEN;
   // timestamp
-  snprintf(buf, UINT64_LEN + 1, "%020" PRIu64, timestamp);
+  snprintf(buf, UINT64_LEN + 1, "%020" PRIu64, ctx->timestamp);
   memcpy(ptr, buf, UINT64_LEN);
   ptr += UINT64_LEN;
   // hmac
-  memcpy(ptr, hmac, TA_AES_HMAC_SIZE);
+  memcpy(ptr, ctx->hmac, TA_AES_HMAC_SIZE);
   ptr += TA_AES_HMAC_SIZE;
   // ciphertext length
   memcpy(ptr, str_ciphertext_len, UINT32_LEN);
   ptr += UINT32_LEN;
   // ciphertext
-  memcpy(ptr, ciphertext, ciphertext_len);
-  *out_msg_len = IV_LEN + UINT64_LEN + TA_AES_HMAC_SIZE + UINT32_LEN + ciphertext_len;
+  memcpy(ptr, ctx->ciphertext, ctx->ciphertext_len);
+  *out_msg_len = IV_LEN + UINT64_LEN + TA_AES_HMAC_SIZE + UINT32_LEN + ctx->ciphertext_len;
 
   return SC_OK;
 }
 
-status_t deserialize_msg(char *msg, const uint8_t *iv, size_t *ciphertext_len, char *ciphertext, uint64_t *timestamp,
-                         uint8_t *hmac) {
+status_t deserialize_msg(const char *msg, ta_cipher_ctx *ctx) {
   /* FIXME: Provide some checks here */
   char str_ciphertext_len[UINT32_LEN + 1] = {};
   char buf[UINT64_LEN + 1] = {0};
-  char *ptr = msg;
+  const char *ptr = msg;
+  if(ptr == NULL){
+    fprintf(stderr,"NULL message\n");
+    return SC_UTILS_TEXT_DESERIALIZE;
+  }
   uint32_t ciphertext_len_tmp;
   // initialize vector
-  memcpy((char *)iv, ptr, IV_LEN);
+  memcpy(ctx->iv, ptr, IV_LEN);
```

Comment:
Do we need to check iv is non-NULL

Reply:
The initialized vector inside cipher_ctx is an array.